src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Resolvers/DynamicGenericResolver.cs

```diff
@@ -313,13 +313,53 @@ internal static object GetFormatter(Type t)
                     return CreateInstance(typeof(GenericDictionaryFormatter<,,>), new[] { keyType, valueType, t });
                 }
 
+                // generic readonly dictionary
+                var readOnlyDictionaryDef = ti.ImplementedInterfaces.FirstOrDefault(x => x.GetTypeInfo().IsConstructedGenericType() && x.GetGenericTypeDefinition() == typeof(IReadOnlyDictionary<,>));
+                if (readOnlyDictionaryDef != null)
+                {
+                    Type keyType = readOnlyDictionaryDef.GenericTypeArguments[0];
+                    Type valueType = readOnlyDictionaryDef.GenericTypeArguments[1];
+                    Type[] allowedParameterTypes = new Type[]
+                    {
+                        typeof(IDictionary<,>).MakeGenericType(keyType, valueType),
+                        typeof(IReadOnlyDictionary<,>).MakeGenericType(keyType, valueType),
+                        typeof(IEnumerable<>).MakeGenericType(typeof(KeyValuePair<,>).MakeGenericType(keyType, valueType)),
+                    };
+                    foreach (var constructor in ti.DeclaredConstructors)
+                    {
+                        ParameterInfo[] parameters = constructor.GetParameters();
+                        if (parameters.Length == 1 &&
+                            allowedParameterTypes.Any(allowedType => allowedType.IsAssignableFrom(parameters[0].ParameterType)))
```

Comment:
I think you have the `IsAssignableFrom` backwards. If the ctor allowed `MyCustomDictionary`, then `IDictionary<K,V>` may be assignable from it, but ultimately our `Activator.CreateInstance` call would fail because our `Dictionary<K,V>` object cannot be assigned to `MyCustomDictionary`

If you swap it to be:

```cs
if (parameters.Length == 1 && parameters[0].ParameterType.IsAssignableFrom(typeof(Dictionary<,>).MakeGenericType(keyType, valueType))
```

Then we know definitively that the ultimate invocation should succeed, and all the 3 types you were previously checking individually will all just work.